### PR TITLE
GEODE-9817: Enable customized source set paths for ClassAnalysisRule

### PR DESCRIPTION
**Pull request by kirklund**
Adds support for pluggable source set paths to ClassAnalysisRule.

PROBLEM

Modules external to Geode must be structured the same as Geode
source code in order to use ClassAnalysisRule and the
Analyze*Serializables tests. This is necessary to better facilitate
pluggability of modules that need to provide sanctioned serializable 
lists.

SOLUTION

Add source set path customization to ClassAnalysisRule, introduce
a new layer of Analyze*Serializables test base classes that can be
directly extended in order to customize source set paths in
ClassAnalysisRule. Also includes improvements to some iterating
of classes during analysis.

### CODE REVIEW

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:352]: This code doesn't seem to be needed when running in IntelliJ. Delete it.

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:55]: This javadoc should now reference the new `AnalyzeSerializablesWithClassAnalysisRuleTestBase`.

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:207]: Delete this line.

File: geode-junit/src/main/java/org/apache/geode/test/junit/rules/ClassAnalysisRule.java
Comment (kirklund) [inline geode-junit/src/main/java/org/apache/geode/test/junit/rules/ClassAnalysisRule.java:118]: Running these tests in IntelliJ uses `build/classes/java`. These special cases for IntelliJ don't seem to be used anymore. Delete them.

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java]: Done

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java]: Done

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java]: Done

File: geode-junit/src/main/java/org/apache/geode/test/junit/rules/ClassAnalysisRule.java
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/test/junit/rules/ClassAnalysisRule.java]: Done

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (jchen21) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:204]: Why there is `NullPointerException` here? Better add some comment here.

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (jchen21) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:230]: Why not catch `CacheClosedException` here, instead of `Exception`?

**Review by dschneider-pivotal — APPROVED** [review]

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeSerializablesTestBase.java
Comment (jchen21) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeSerializablesTestBase.java:370]: Better have consistent error message with line 361.

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (nabarunnag) [inline geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:231]: No hate :) but I think we can put the message in the sys out below mentioning that geode-membership cannot load this class.

**Review by nabarunnag — APPROVED** [review]

**Review by agingade — APPROVED** [review]
I am not expert in this area; but did not find any noticeable issues with the PR.

**Review by DonalEvans — APPROVED** [review]
Approving for the files I'm code owner for.

**Review by upthewaterspout — APPROVED** [review]
Haven't done a full review but the files I'm codeowner for look fine.

**Review by kamilla1201 — APPROVED** [review]

**Review by mhansonp — APPROVED** [review]

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeDataSerializablesTestBase.java:230]: AnalyzeMembershipSerializablesIntegrationTest extends AnalyzeDataSerializablesTestBase and lives in geode-membership which does not have a dependency on geode-core. So when that test runs, it was throwing ClassNotFoundException for CacheClosedException. Then I removed ClassNotFoundException and changed it to Exception with the dreaded type check.

File: geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeSerializablesTestBase.java
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/codeAnalysis/AnalyzeSerializablesTestBase.java:370]: I changed this line from `throw new AssertionError` to `errorCollector.addError(new AssertionError)`. Probably not a big deal since it's a test, but I can change this. I was holding off from doing much more work on these TestBase classes, but I'd like to rework them in the near future.

**Review by jchen21 — APPROVED** [review]

**Review by jinmeiliao — APPROVED** [review]

**Comment by kirklund** [comment]
distributed-test-openjdk11 was in MemberMXBeanDistributedTest > testBucketCount which is GEODE-8191.

**Pull request merged**
